### PR TITLE
adapter: fix AS OF AT LEAST timestamp selection

### LINKED ISSUE #19092: AS OF AT LEAST doesn't work as expected
### What version of Materialize are you using?

f72e00d022e2d43b0af0b610ebf24032d2864caa

### What is the issue?

After https://github.com/MaterializeInc/materialize/pull/18986 landed, I still see the `AS OF` timestamp that's selected being very recent, even when there are earlier timestamps available.

Here's the error I see when trying to query historical cluster utilization:
```
SUBSCRIBE (SELECT r.id,  u.cpu_percent,  u.memory_percent
FROM mz_cluster_replicas r
JOIN mz_internal.mz_cluster_replica_utilization u ON u.replica_id = r.id
WHERE r.cluster_id = 's2'
)  AS OF AT LEAST TIMESTAMP '2023-05-02T20:50:10.633Z'  UP TO TIMESTAMP '2023-05-02T21:05:10.633Z';
ERROR:  subscription lower ("as of") bound is beyond its upper ("up to") bound: 1683061510633 < 1683061998874
```

Here is the explain for that query:
```
explain timestamp for SELECT r.id,
  u.cpu_percent,
  u.memory_percent
FROM mz_cluster_replicas r
JOIN mz_internal.mz_cluster_replica_utilization u ON u.replica_id = r.id
WHERE r.cluster_id = 'u1';
                                 Timestamp
---------------------------------------------------------------------------
                 query timestamp: 1683062251906 (2023-05-02 21:17:31.906) +
           oracle read timestamp: 1683062251906 (2023-05-02 21:17:31.906) +
 largest not in advance of upper: 1683062251906 (2023-05-02 21:17:31.906) +
                           upper:[1683062251907 (2023-05-02 21:17:31.907)]+
                           since:[1680470251000 (2023-04-02 21:17:31.000)]+
         can respond immediately: true                                    +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source mz_catalog.mz_cluster_replicas (s330, storage):                   +
                   read frontier:[1680470251000 (2023-04-02 21:17:31.000)]+
                  write frontier:[1683062251907 (2023-05-02 21:17:31.907)]+
                                                                          +
 source mz_internal.mz_cluster_replica_metrics (s332, storage):           +
                   read frontier:[1680470251000 (2023-04-02 21:17:31.000)]+
                  write frontier:[1683062251907 (2023-05-02 21:17:31.907)]+
                                                                          +
 source mz_internal.mz_cluster_replica_sizes (s333, storage):             +
                   read frontier:[1680470251000 (2023-04-02 21:17:31.000)]+
                  write frontier:[1683062251907 (2023-05-02 21:17:31.907)]+

(1 row)
```

### PR DESCRIPTION
**Pull request by mjibson**
Add a test harness for timestamp selection. Use it to identify and fix a bug in AT LEAST which was using the oracle read ts.

Fixes #19092 

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/adapter/tests/timestamp_selection.rs [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+use mz_expr::MirScalarExpr;
+use mz_repr::{Datum, GlobalId, ScalarType, Timestamp};
+use mz_sql::plan::QueryWhen;
+use mz_sql_parser::ast::TransactionIsolationLevel;
+use mz_storage_client::types::sources::Timeline;
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(transparent)]
+struct Set {
+    ids: BTreeMap<String, Vec<Timestamp>>,
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/tests/timestamp_selection.rs:107]: Or even a struct with named fields would make the tests easier to read.

File: src/adapter/src/coord/timestamp_selection.rs [hunk 2]
```diff
@@ -316,45 +384,104 @@ impl Coordinator {
...
+        let invalid = invalid_indexes
+            .into_iter()
+            .chain(invalid_sources)
+            .collect::<Vec<_>>();
+        format!(
+            "Timestamp ({}) is not valid for all inputs: {:?}",
+            candidate, invalid,
+        )
+    }
+}
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/coord/timestamp_selection.rs:443]: All code movement.

File: src/adapter/src/coord/timestamp_selection.rs [hunk 3]
```diff
@@ -104,16 +106,109 @@ impl<T: TimestampManipulation> TimestampContext<T> {
...
+        &self
+            .controller
+            .storage
+            .collection(id)
+            .expect("id does not exist")
+            .write_frontier
+    }
+}
+
+pub trait TimestampProvider {
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/coord/timestamp_selection.rs:179]: Same! There are methods on the controllers that explain the difference. I've copied the first sentence of those docs to the trait.

File: src/adapter/src/coord/timestamp_selection.rs [hunk 4]
```diff
@@ -264,48 +359,21 @@ impl Coordinator {
...
-            .determine_timestamp
-            .with_label_values(&[
-                match det.respond_immediately() {
-                    true => "true",
-                    false => "false",
-                },
-                isolation_level.as_str(),
-                &compute_instance.to_string(),
-            ])
-            .inc();
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/coord/timestamp_selection.rs:278]: They are still on the `determine_timestamp` method on Coord.

File: src/adapter/tests/timestamp_selection.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        self.compute.get(&(instance, id)).unwrap().read.borrow()
+    }
+
+    fn compute_read_capability<'a>(
+        &'a self,
+        instance: ComputeInstanceId,
+        id: GlobalId,
+    ) -> &'a timely::progress::Antichain<Timestamp> {
+        &self.compute.get(&(instance, id)).unwrap().read
+    }
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/tests/timestamp_selection.rs:169]: No idea, but it's sufficient for the tests. If we find a test case that cares about this we can split it up.

File: src/adapter/tests/timestamp_selection.rs [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+// Test determine_timestamp.
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/tests/timestamp_selection.rs:87]: I added more doc comments to the test function itself that should be useful enough to understand what's going on, even if they aren't fully documented.

File: src/adapter/tests/timestamp_selection.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/tests/timestamp_selection.rs:107]: Changed to a struct with named fields.

**Review by jkosh44 — APPROVED** [review]
LGTM, awesome that we have a test framework for this now!

**Review by def- — COMMENTED** [review]
Yay for testing improvements! The `AS OF AT LEAST` sounds useful for testing too compared to our TD approach of waiting around a few seconds.

**Pull request merged**